src/Types.jl

```diff
@@ -900,63 +901,260 @@ function ensure_resolved(env::EnvCache,
+                    pkgerror("multiple registries with name `$(reg.name)`, please specify with uuid.")
+                end
+                reg.url = known.url
+            end
+        end
+    end
+end
+
+clone_or_cp_registries(regs::Vector{RegistrySpec}, depot::String=depots1()) =
+    clone_or_cp_registries(Context(), regs, depot)
+function clone_or_cp_registries(ctx::Context, regs::Vector{RegistrySpec}, depot::String=depots1())
+    populate_known_registries_with_urls!(regs)
+    # TODO Bail early on preview
+    for reg in regs
+        if reg.path !== nothing && reg.url !== nothing
+            pkgerror("ambiguous registry specification; both url and path is set.")
+        end
+        # clone to tmpdir first
+        tmp = mktempdir()
+        if reg.path !== nothing # copy from local source
+            printpkgstyle(stdout, :Copying, "registry from `$(Base.contractuser(reg.path))`")
+            cp(reg.path, tmp; force=true)
+        elseif reg.url !== nothing # clone from url
             Base.shred!(LibGit2.CachedCredentials()) do creds
-                printpkgstyle(stdout, :Cloning, "default registries into $user_regs")
-                for (reg, url) in DEFAULT_REGISTRIES
-                    path = joinpath(user_regs, reg)
-                    LibGit2.with(GitTools.clone(url, path; header = "registry $reg from $(repr(url))", credentials = creds)) do repo
-                    end
+                LibGit2.with(GitTools.clone(reg.url, tmp; header = "registry from $(repr(reg.url))",
+                    credentials = creds)) do repo
                 end
             end
+        else
+            pkgerror("no path or url specified for registry")
+        end
+        # verify that the clone looks like a registry
+        if !isfile(joinpath(tmp, "Registry.toml"))
+            pkgerror("no `Registry.toml` file in cloned registry.")
+        end
+        registry = read_registry(joinpath(tmp, "Registry.toml"); cache=false) # don't cache this tmp registry
+        verify_registry(registry)
+        # copy to `depot`
+        # slug = Base.package_slug(UUID(registry["uuid"]))
+        regpath = joinpath(depot, "registries", registry["name"]#=, slug=#)
+        ispath(dirname(regpath)) || mkpath(dirname(regpath))
+        if isdir_windows_workaround(regpath)
+            existing_registry = read_registry(joinpath(regpath, "Registry.toml"))
+            if registry["uuid"] == existing_registry["uuid"]
+                @info("registry `$(registry["name"])` already exist in `$(Base.contractuser(regpath))`.")
+            else
+                throw(PkgError("registry `$(registry["name"])=\"$(registry["uuid"])\"` conflicts with " *
+                    "existing registry `$(existing_registry["name"])=\"$(existing_registry["uuid"])\"`. " *
+                    "To install it you can clone it manually into " *
+                    "`$(Base.contractuser(joinpath(depot, "registries", registry["name"]*"_2")))` or similar."))
```

Comment:
I have to say that it feels a bit shitty to leave this so close to working. This is literally the *only* problem with adding another registry with the same name. I am going to change this to a warning and then we clone it into `registries/name-uuid` in case of conflicts.

Reply:
Ok! Would you rename the existing one to use the same scheme then? I would avoid using the whole UUID since long paths cause problems on Windows at least—use the slug instead.

Reply:
We can rename it too I guess.
Just adding a note to self; when a user adds a third registry with the same name it is not enough to just check the path, we need instead to check that there is no registry with the same name in the depot.

Reply:
Of course it is 🙂 Just annoying that there is one single spot in the code that can not cope with multiple ones like this.

Reply:
FWIW I still vote for just erroring out and keeping it simple. I think filesystem-as-a-database is a nice default

Reply:
Since we are relying on GitHub so much already... Perhaps `registries/UserName/RegName` is enough to avoid name collisions?

Reply:
Not a bad idea—we could support GitLab and others in a similar fashion.

Reply:
I think I will leave it like this for now then.

> Perhaps `registries/UserName/RegName` is enough to avoid name collisions?

Yea I guess, but that does not solve the problem raised in the other thread (https://github.com/JuliaLang/Pkg.jl/issues/812#issuecomment-429187763 and https://github.com/JuliaLang/Pkg.jl/issues/812#issuecomment-429484241) where the user wanted to "just make a clone".

Reply:
I like this plan, seems simpler and easier to work with than slugs or whatever else.